crates/paralegal-policy/src/context.rs

```diff
@@ -462,45 +462,70 @@ impl Context {
     /// always return the same result for the same input.
     pub fn always_happens_before<'a>(
         &self,
-        ctrl: ControllerId,
         starting_points: impl Iterator<Item = Node<'a>>,
         mut is_checkpoint: impl FnMut(Node) -> bool,
         mut is_terminal: impl FnMut(Node) -> bool,
     ) -> Result<AlwaysHappensBefore> {
-        let mut seen = HashSet::<Node>::new();
+        let mut seen = HashSet::<&DataSink>::new();
         let mut num_reached = 0;
         let mut num_checkpointed = 0;
 
-        let mut queue = starting_points.collect::<Vec<_>>();
+        let mut queue = starting_points
+            .filter_map(|n| match n.typ.as_data_source() {
+                Some(ds) => Some((
+                    ds,
+                    n.ctrl_id,
+                    &self.desc().controllers.get(&n.ctrl_id).unwrap().data_flow.0,
+                )),
+                None => {
+                    assert_warning!(
+                        *self,
+                        false,
+                        format!(
+                            "found starting point {:?} that cannot be converted to a datasource",
+                            n
+                        )
+                    );
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
 
         let started_with = queue.len();
 
-        let flow = &self
-            .desc()
-            .controllers
-            .get(&ctrl)
-            .ok_or_else(|| anyhow!("Controller not found"))?
-            .data_flow
-            .0;
-
-        while let Some(current) = queue.pop() {
-            if is_checkpoint(current) {
+        // Return whether node is checkpoint or terminal and increment those respective counters
+        let mut check_node = |node: Node| -> bool {
```

Comment:
Nice touch moving the checks into this central location. In fact I had thought about suggesting this but didn't want to micro-manage 😄